src/app/content/utils/canonicalUrl.ts

```diff
@@ -38,34 +38,36 @@ export async function getCanonicalUrlParams(
         getBookVersionFromUUIDSync(id),
         `We've already filtered out books that are not in the BOOK configuration`
       ).defaultVersion;
-      const useCurrentBookAsCanonical = book.id === id  && hasOSWebData(book);
-      canonicalBook = useCurrentBookAsCanonical ? book : await getBook(id, version);
-      canonicalPageId = CANONICAL_PAGES_MAP[pageId] || canonicalPageId;
-      treeSection = findArchiveTreeNodeById(canonicalBook.tree, canonicalPageId);
-
-      if (!useCurrentBookAsCanonical) {
-        const newMap = getCanonicalMap(canonicalBook.id);
-        // stop when we run out of canonical maps to check
-        done = !newMap.length || isEqual(canonicalMap, newMap);
-        // throw if the new map has already been checked
-        if (!done && mapsChecked.find((map) => isEqual(map, newMap))) {
-          throw new Error(`Loop encountered in map for ${canonicalBook.id}`);
-        }
-        canonicalMap = newMap;
+      canonicalBook = book.id === id  && hasOSWebData(book) ? book : await getBook(id, version);
+      canonicalPageId = CANONICAL_PAGES_MAP[canonicalPageId] || canonicalPageId;
+      const treeSection = findArchiveTreeNodeById(canonicalBook.tree, canonicalPageId);
+      // use the most recent canonical page found if none is found in current canonical book
+      if (!treeSection && canonicalBookWithPage) {
+        done = true;
         break;
+      } else if (treeSection) {
+        canonicalBookWithPage = {canonicalBook, treeSection};
+      }
+
+      // check if canonical book maps to another book
+      const newMap = getCanonicalMap(canonicalBook.id);
+      done = !newMap.length || isEqual(canonicalMap, newMap);
+      // throw if the new map has already been checked
+      if (!done && mapsChecked.find((map) => isEqual(map, newMap))) {
+        throw new Error(`Loop encountered in map for ${canonicalBook.id}`);
       }
-      done = true;
+      canonicalMap = newMap;
     }
   }
 
-  // use current page if page not found in canonical book
-  if (!treeSection && canonicalMap.length) {
-    treeSection = findArchiveTreeNodeById(book.tree, pageId);
-    canonicalBook = book;
+  // use current page as canonical if page not found in canonical book
+  if (!canonicalBookWithPage && canonicalMap.length) {
```

Comment:
This isn't a new change, but what is the canonicalMap existence check here a signal for? I'm curious if these two ifs can be collapsed but I'm not sure what the check does.

Reply:
Historically this function only generated a canonical URL if a canonical page was successfully found. If the book wasn't found in the config (see [line 24](https://github.com/openstax/rex-web/pull/1644/files#diff-a7b7fc51715f392a83e1785cc3c82834c33371291b07eebced8573f006f25d21R24)) OR if the book was found but a canonical page was not, we returned null instead. Now we've decided any page that exists should get a canonical URL (if there's no canonical page it just references its own URL). So line 64 handles that case and line 68 handles the case where an actual `canonicalBookWithPage` was found.

It's possible that we never want to return null and just want to return a URL or throw an error instead. In that case I think we could remove the `canonicalMap.length` check.